src/Sentry.Unity.Android/JniExecutor.cs

```diff
@@ -50,74 +61,110 @@ private void DoWork()
             catch (Exception e)
             {
-                Debug.unityLogger.Log(LogType.Exception, UnityLogger.LogTag, $"Error during JNI execution: {e}");
+                _logger?.LogError(e, "Error during JNI execution.");
+                _taskCompletionSource?.SetException(e);
             }
         }
 
         AndroidJNI.DetachCurrentThread();
     }
 
-    public TResult? Run<TResult>(Func<TResult?> jniOperation)
+    public TResult? Run<TResult>(Func<TResult?> jniOperation, TimeSpan? timeout = null)
     {
         lock (_lock)
         {
+            timeout ??= DefaultTimeout;
+            using var timeoutCts = new CancellationTokenSource(timeout.Value);
             _taskCompletionSource = new TaskCompletionSource<object?>();
             _currentTask = jniOperation;
             _taskEvent.Set();
 
             try
             {
-                return (TResult?)_taskCompletionSource.Task.GetAwaiter().GetResult();
+                _taskCompletionSource.Task.Wait(timeoutCts.Token);
+                return (TResult?)_taskCompletionSource.Task.Result;
+            }
+            catch (OperationCanceledException)
+            {
+                _logger?.LogError("JNI execution timed out.");
+                return default;
             }
             catch (Exception e)
             {
-                Debug.unityLogger.Log(LogType.Exception, UnityLogger.LogTag, $"Error during JNI execution: {e}");
+                _logger?.LogError(e, "Error during JNI execution.");
+                return default;
+            }
+            finally
+            {
+                _currentTask = null!;
             }
-
-            return default;
         }
     }
 
-    public void Run(Action jniOperation)
+    public void Run(Action jniOperation, TimeSpan? timeout = null)
     {
         lock (_lock)
         {
+            timeout ??= DefaultTimeout;
+            using var timeoutCts = new CancellationTokenSource(timeout.Value);
             _taskCompletionSource = new TaskCompletionSource<object?>();
             _currentTask = jniOperation;
             _taskEvent.Set();
 
             try
             {
-                _taskCompletionSource.Task.Wait();
+                _taskCompletionSource.Task.Wait(timeoutCts.Token);
+            }
+            catch (OperationCanceledException)
+            {
+                _logger?.LogError("JNI execution timed out.");
             }
             catch (Exception e)
             {
-                Debug.unityLogger.Log(LogType.Exception, UnityLogger.LogTag, $"Error during JNI execution: {e}");
+                _logger?.LogError(e, "Error during JNI execution.");
+            }
+            finally
+            {
+                _currentTask = null!;
             }
         }
     }
 
     public void Dispose()
     {
+        if (_isDisposed)
+        {
+            return;
+        }
+
+        _isDisposed = true;
+
         _shutdownSource.Cancel();
+        _workerThread?.Join(100);
```

Comment:
It can throw for:

Exceptions
[ThreadStateException](https://learn.microsoft.com/en-us/dotnet/api/system.threading.threadstateexception?view=net-9.0)
The caller attempted to join a thread that is in the [Unstarted](https://learn.microsoft.com/en-us/dotnet/api/system.threading.threadstate?view=net-9.0#system-threading-threadstate-unstarted) state.

[ThreadInterruptedException](https://learn.microsoft.com/en-us/dotnet/api/system.threading.threadinterruptedexception?view=net-9.0)
The thread is interrupted while waiting.


We should catch this here and suppress it

Reply:
Also worth reading the remarks of these APIs.

Curious why we use Thread directly instead of the Task abstraction. It's by default a background thread and it's much better to deal with cancellation, error handling etc (something for a potential future task, to change this)

Reply:
> Curious why we use Thread directly instead of the Task abstraction.

The executor needs to run all the calls on a single thread. If we converted this to a task, we'd still need to ensure that so no async/await, etc. In that case, I only see risks that we do something wrong here now or in the future version of the code which would lead to the task being rescheduled to a different thread in the pool.